### ISSUE TITLE
fastdp compatibility with 1.1

### ISSUE THREAD

**Issue by dpw** [weaveworks/weave#1448]
Following on from #1396 and #1438:

There's an easy part to this, and a hard part:

The easy part is to making sure that when a version 1.1 router and a #1438 router connect, they fall back to the sleeve Overlay.  #1438 does not do that, but it is trivial to make it do so.

The hard part is fast datapath's reliance on peer short ids. Version 1.1 peers will not generate short ids for themselves, and will not relay short id information.  Fast datapath cannot forward to peers it does not have a short id for.  So it's possible to envisage pathological cases where two groups of version 1.2 peers are connected only through a version 1.1 peer, leading to an effective partition in the overlay network.

(There is also an issue that GOB zeroes out fields when deserializing an old version of a struct.  So the range of `Peer.ShortID` will have to be adjusted from [0, 4095] to [1, 4096] in order to avoid conflating a missing short id with the valid short id of zero.)

I can think of two approaches to supporting rolling upgrades:

The first is to require users to upgrade to 1.2 via an intermediate step, where they start 1.2 in a fastdp disabled mode.  I plan such a mode anyway, so no extra development work will be required.  Once a cluster has been upgraded to 1.2 with fastdp disabled, the user is free to restart routers with fastdp enabled.

The downside of this approach is the potential for confusion if users try to upgrade without following the release notes carefully.  With 1.1, we have seen that debugging issues that might or might not be compatibility bugs is a time consuming process.  Given that this approach means that we genuinely will have a compatibility issue unless the appropriate option is passed, I am concerned about what will ensue.

The second option is to run the fastdp and sleeve overlays in parallel, and use the sleeve overlay whenever we need to forward a packet to (and from) a peer for which we lack a short id.  Currently in #1438, `OverlaySwitch` does not do this:  When they fastdp `OverlayForwarder` confirms, the sleeve `OverlayForwarder` gets stopped.  In the long term, that behaviour is desirable: We don't want to maintain the sleeve forwarder machinery, sending heartbeats and fragmentation tests etc., if it is not needed.  But it would be fairly easy to change that for 1.2.  Then in 1.3, we could change it back.

**Comment by awh** [comment]
> The downside of this approach is the potential for confusion ... we genuinely will have a compatibility issue unless the appropriate option is passed

What if fastdp is disabled by default?

**Comment by awh** [comment]
> Version 1.1 peers ... will not relay short id information

Is there anything we can do to future proof ourselves against a similar scenario in the future e.g. have peers ignore attributes they don't understand but still relay them on?

**Comment by dpw** [comment]
> What if fastdp is disabled by default?

It would seem strange not to enable it in 1.2 just for the sake of this issue.  Hey, why not just merge the short ids change for 1.2, and put the rest of fastdp back on the shelf for another couple of months?

**Comment by dpw** [comment]
> Is there anything we can do to defend ourselves against a similar scenario in the future e.g. have peers ignore attributes they don't understand but still relay them on?

"ignore but propagate" behaviour for peer attributes would not substantially change the situation:  We would still have the issue that 1.1 peers do not have a short id.  And it's unclear to me that that behaviour would help in general with compatibility issues.

The question of how you achieve systematic support for rolling upgrades in non-trivial distributed systems is an interesting one. I think there are pragmatic solutions, but I'm not sure it's feasible to get there from where we are now with Weave Net.

**Comment by awh** [comment]
>  Hey, why not just merge the short ids change for 1.2, and put the rest of fastdp back on the shelf for another couple of months?

I understand you are frustrated, but the implication that I am unconcerned with getting fastdp into a release seems misplaced given that I volunteered for review duty in order to do exactly that :smile:

The suggestion was made in good faith - other major (potentially disruptive) components such as weaveDNS and the proxy have spent one or more release cycles disabled by default but available for advanced users, so its entirely in-line with our established process; putting fastdp 'back on the shelf' would not obtain the same benefit, and is clearly not what I meant.

That being said, if you are equally concerned 'about what will ensue' as you are about having *everyone* cut over to fastdp in the instant they upgrade to 1.2 then option 2 is the only choice would you not say?

**Comment by awh** [comment]
> So it's possible to envisage pathological cases where two groups of version 1.2 peers are connected only through a version 1.1 peer, leading to an effective partition in the overlay network.

This is a transient condition though, is it not, as long as we recommend against prolonged running of mixed 1.1/1.2 networks? The rolling upgrade is going to result in sporadic losses of connectivity anyway, so is this any worse as long as it doesn't persist?

**Comment by awh** [comment]
> "ignore but propagate" behaviour for peer attributes would not substantially change the situation

I think I have misunderstood the reason for this:

> Fast datapath cannot forward to peers it does not have a short id for

More thought required!

**Comment by awh** [comment]
> Fast datapath cannot forward to peers it does not have a short id for

Do you really mean 'via' here and not 'to'?

**Comment by dpw** [comment]
> I understand you are frustrated, but the implication that I am unconcerned with getting fastdp into a release seems misplaced given that I volunteered for review duty in order to do exactly that :sw mile:

I'm not at all frustrated.  I did not intend to imply that you or anyone else is unconcerned about getting fastdp into a release.  Putting out a release with fastdp disabled by default *due to this issue* seems like a strange decision to me, given that rolling upgrades are of importance to relatively few users (albeit ones we want to keep happy).

> The suggestion was made in good faith - other major (potentially disruptive) components such as weaveDNS and the proxy have spent one or more release cycles disabled by default but available for advanced users, so its entirely in-line with our established process; putting fastdp 'back on the shelf' would not obtain the same benefit, and is clearly not what I meant.

I anticipated that argument would crop up sooner or later.  But it is unrelated to the issue of rolling upgrades, so I'm not sure this is the place for it.

> That being said, if you are equally concerned 'about what will ensue' as you are about having everyone cut over to fastdp in the instant they upgrade to 1.2 then option 2 is the only choice would you not say?

I'm not concerned about having fastdp enabled by default in 1.2.  My goal is that we *could* release 1.2 with fastdp enabled, and that still seems viable to me.  But I could change my mind on that.  And perhaps it shouldn't be my decision to make.

Maybe I should have spelled it out more, but I think option 2 is distinctly preferable to option 1.  I.e. "here's the really easy thing to do, but it sucks, and here's the much better thing to do, which isn't a whole lot of work either."

**Comment by dpw** [comment]
> > Fast datapath cannot forward to peers it does not have a short id for
> 
> Do you really mean 'via' here and not 'to'?

Neither really.

More precisely:  Fast datapath cannot forward a packet with a source or destination peer for which it does not know the short id.

So version 1.1 peers present two problems: First, they do not have short ids, and second, they can even hide the short ids of other version 1.2 peers.

In practice, it is probably hard to exhibit those issues for small and/or fully connected weave networks: If a 1.2 peer is forwarding to a 1.1 peer, the short id of the destination peer will not be known, but it won't be using fastdp to forward the packet anyway so it doesn't matter.

**Comment by dpw** [comment]
> > So it's possible to envisage pathological cases where two groups of version 1.2 peers are connected only through a version 1.1 peer, leading to an effective partition in the overlay network.
> 
> This is a transient condition though, is it not, as long as we recommend against prolonged running of mixed 1.1/1.2 networks? The rolling upgrade is going to result in sporadic losses of connectivity anyway, so is this any worse as long as it doesn't persist?

If you have a topology (with appropriately placed 1.1 peers) it will persist until the 1.1 peers are removed.  I think users understand that doing `weave-1.1 stop-router && weave-1.2 launch-router` leads to transient loss of connectivity to the weave network.  But the natural expectation for a rolling upgrade is that it does not impact things beyond that.

Though thinking through the details now, there's something I missed before.  If you do `weave-1.1 stop-router && weave-1.2 launch-router`, the 1.2 router will be running with a `weave` bridge netdev, not an ODP datapath netdev.  That causes it to disable fastdp.  So option 1 happens automagically.

It's still not ideal though.  If you add new 1.2 peers, you can get the kinds of issues alluded to above.

**Comment by awh** [comment]
> Fast datapath cannot forward a packet with a source or destination peer for which it does not know the short id.

OK, so I think I did understand; given the minimal initial steps you outlined

> when a version 1.1 router and a #1438 router connect, they fall back to the sleeve Overlay

in the following network:

````
A(1.2) <--fdp--> B(1.2) <-- sleeve --> C(1.1) <-- sleeve --> D(1.2)
````

A cannot forward a packet to D via its fastdp link to B because it doesn't have D's short ID. This is what motivated my thought around 'ignore but propagate'; if A could learn D's short peer ID via C, it would be able to forward to B, which could then forward to C because

>  If a 1.2 peer is forwarding to a 1.1 peer, the short id of the destination peer will not be known, but it won't be using fastdp to forward the packet anyway so it doesn't matter.

Have I got that right?

I agree this benefit of 'ignore but propagate' may not generalise well to other use cases and therefore may not be worth the cost of implementation.

**Comment by awh** [comment]
Actually even if what I've written above is true, it would still be broken becuase A could never forward to C as it would still lack the short peer ID.

**Comment by dpw** [comment]
> Have I got that right?

Yes.  Your diagram is exactly the same one I drew on the white board for Matthias after you left the office.

> Actually even if what I've written above is true, it would still be broken because A could never forward to C as it would still lack the short peer ID.

Yes.  So 'ignore but propagate' is not enough.  While ever there are version 1.1 peers around, we need to have sleeve forwarders everywhere.  That's why I think doing better than option 2 is tricky.

**Comment by rade** [comment]
Current plan is to not do anything here and instead instruct users in release notes to first upgrade to 1.2 with fdp disabled, and then, if they want fdp, do a (rolling, if they like) restart of all nodes with fdp enabled.

**Comment by dpw** [comment]
> The second option is to run the fastdp and sleeve overlays in parallel, and use the sleeve overlay whenever we need to forward a packet to (and from) a peer for which we lack a short id. Currently in #1438, OverlaySwitch does not do this: When they fastdp OverlayForwarder confirms, the sleeve OverlayForwarder gets stopped. In the long term, that behaviour is desirable: We don't want to maintain the sleeve forwarder machinery, sending heartbeats and fragmentation tests etc., if it is not needed. But it would be fairly easy to change that for 1.2. Then in 1.3, we could change it back.

This proposal is implemented in #1522.

**Comment by rade** [comment]
Fixed by #1522.
